{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Import modules"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import spacy\n",
    "import pandas as pd\n",
    "import os\n",
    "import numpy as np \n",
    "import matplotlib.pyplot as plt\n",
    "from sklearn.metrics.pairwise import cosine_distances\n",
    "from sklearn.metrics.pairwise import euclidean_distances\n",
    "from sklearn.metrics.pairwise import manhattan_distances\n",
    "from pyod.models import ocsvm, iforest, sos  \n",
    "import scipy"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Create UDFs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "def calc_distance(df_vectors, df_vector_averages):\n",
    "    \n",
    "    # Cosine distance\n",
    "    cos_dist_matrix = cosine_distances(df_vectors, df_vector_averages)\n",
    "    cos_dist = pd.Series(np.diagonal(cos_dist_matrix), name='cos_dist')\n",
    "\n",
    "    # Euclidian distance \n",
    "    euc_dist_matrix = euclidean_distances(df_vectors, df_vector_averages)\n",
    "    euc_dist = pd.Series(np.diagonal(euc_dist_matrix), name='euc_dist')\n",
    "\n",
    "    # Manhattan distance \n",
    "    manh_dist_matrix = manhattan_distances(df_vectors, df_vector_averages)\n",
    "    manh_dist = pd.Series(np.diagonal(manh_dist_matrix), name='manh_dist')\n",
    "    \n",
    "    # Build & return df\n",
    "    df_dist = pd.concat([cos_dist, euc_dist, manh_dist], axis=1)\n",
    "    return df_dist\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "def calc_anormality(df_vectors):\n",
    "\n",
    "    # Isolation forest \n",
    "    iforest_ad = iforest.IForest()\n",
    "    iforest_fitted = iforest_ad.fit(df_vectors)\n",
    "    iforest_scores = pd.Series(iforest_fitted.decision_scores_, name = 'iforest_ad')\n",
    "\n",
    "    # One class SVM\n",
    "    ocsvm_ad = ocsvm.OCSVM()\n",
    "    ocsvm_fitted = ocsvm_ad.fit(df_vectors)\n",
    "    ocsvm_scores = pd.Series(ocsvm_fitted.decision_scores_, name = 'ocsvm_ad')\n",
    "\n",
    "    # Stochastic outlier selection SVM\n",
    "    sos_ad = sos.SOS()\n",
    "    sos_fitted = sos_ad.fit(df_vectors)\n",
    "    sos_scores = pd.Series(sos_fitted.decision_scores_, name = 'sos_ad')\n",
    "    \n",
    "    # Build & return df\n",
    "    df_anorm = pd.concat([iforest_scores, ocsvm_scores, sos_scores], axis=1)\n",
    "    return df_anorm\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Generate company word vectors & distance scores from word embeddings"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "dir_path = os.path.normpath(r'C:\\Users\\heinr\\OneDrive\\Desktop\\LARGE DATA\\LinkedIn\\company_level_individual_profiles')\n",
    "dir_path_we = os.path.normpath(r'C:\\Users\\heinr\\OneDrive\\Desktop\\LARGE DATA\\LinkedIn\\company_level_individual_profiles_word_embeddings')\n",
    "path_dir_we_dist = os.path.normpath(r'C:\\Users\\heinr\\OneDrive\\Desktop\\LARGE DATA\\LinkedIn\\company_level_individual_profiles_word_embeddings_dist')\n",
    "path_dir_dist_only = os.path.normpath(r'C:\\Users\\heinr\\OneDrive\\Desktop\\LARGE DATA\\LinkedIn\\company_level_individual_profiles_we_dist_only')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\heinr\\anaconda3\\envs\\linkedin\\lib\\site-packages\\pyod\\models\\sos.py:202: RuntimeWarning: overflow encountered in multiply\n",
      "  beta[i] = beta[i] * 2.0\n",
      "C:\\Users\\heinr\\anaconda3\\envs\\linkedin\\lib\\site-packages\\pyod\\models\\base.py:365: RuntimeWarning: invalid value encountered in greater\n",
      "  self.labels_ = (self.decision_scores_ > self.threshold_).astype(\n",
      "C:\\Users\\heinr\\anaconda3\\envs\\linkedin\\lib\\site-packages\\pyod\\models\\sos.py:202: RuntimeWarning: overflow encountered in multiply\n",
      "  beta[i] = beta[i] * 2.0\n",
      "C:\\Users\\heinr\\anaconda3\\envs\\linkedin\\lib\\site-packages\\pyod\\models\\base.py:365: RuntimeWarning: invalid value encountered in greater\n",
      "  self.labels_ = (self.decision_scores_ > self.threshold_).astype(\n",
      "C:\\Users\\heinr\\anaconda3\\envs\\linkedin\\lib\\site-packages\\pyod\\models\\sos.py:202: RuntimeWarning: overflow encountered in multiply\n",
      "  beta[i] = beta[i] * 2.0\n",
      "C:\\Users\\heinr\\anaconda3\\envs\\linkedin\\lib\\site-packages\\pyod\\models\\base.py:365: RuntimeWarning: invalid value encountered in greater\n",
      "  self.labels_ = (self.decision_scores_ > self.threshold_).astype(\n",
      "C:\\Users\\heinr\\anaconda3\\envs\\linkedin\\lib\\site-packages\\pyod\\models\\sos.py:202: RuntimeWarning: overflow encountered in multiply\n",
      "  beta[i] = beta[i] * 2.0\n",
      "C:\\Users\\heinr\\anaconda3\\envs\\linkedin\\lib\\site-packages\\pyod\\models\\base.py:365: RuntimeWarning: invalid value encountered in greater\n",
      "  self.labels_ = (self.decision_scores_ > self.threshold_).astype(\n",
      "C:\\Users\\heinr\\anaconda3\\envs\\linkedin\\lib\\site-packages\\pyod\\models\\sos.py:202: RuntimeWarning: overflow encountered in multiply\n",
      "  beta[i] = beta[i] * 2.0\n",
      "C:\\Users\\heinr\\anaconda3\\envs\\linkedin\\lib\\site-packages\\pyod\\models\\base.py:365: RuntimeWarning: invalid value encountered in greater\n",
      "  self.labels_ = (self.decision_scores_ > self.threshold_).astype(\n",
      "C:\\Users\\heinr\\anaconda3\\envs\\linkedin\\lib\\site-packages\\pyod\\models\\sos.py:202: RuntimeWarning: overflow encountered in multiply\n",
      "  beta[i] = beta[i] * 2.0\n",
      "C:\\Users\\heinr\\anaconda3\\envs\\linkedin\\lib\\site-packages\\pyod\\models\\base.py:365: RuntimeWarning: invalid value encountered in greater\n",
      "  self.labels_ = (self.decision_scores_ > self.threshold_).astype(\n",
      "C:\\Users\\heinr\\anaconda3\\envs\\linkedin\\lib\\site-packages\\pyod\\models\\sos.py:202: RuntimeWarning: overflow encountered in multiply\n",
      "  beta[i] = beta[i] * 2.0\n",
      "C:\\Users\\heinr\\anaconda3\\envs\\linkedin\\lib\\site-packages\\pyod\\models\\base.py:365: RuntimeWarning: invalid value encountered in greater\n",
      "  self.labels_ = (self.decision_scores_ > self.threshold_).astype(\n",
      "C:\\Users\\heinr\\anaconda3\\envs\\linkedin\\lib\\site-packages\\pyod\\models\\sos.py:202: RuntimeWarning: overflow encountered in multiply\n",
      "  beta[i] = beta[i] * 2.0\n",
      "C:\\Users\\heinr\\anaconda3\\envs\\linkedin\\lib\\site-packages\\pyod\\models\\base.py:365: RuntimeWarning: invalid value encountered in greater\n",
      "  self.labels_ = (self.decision_scores_ > self.threshold_).astype(\n",
      "C:\\Users\\heinr\\anaconda3\\envs\\linkedin\\lib\\site-packages\\pyod\\models\\sos.py:202: RuntimeWarning: overflow encountered in multiply\n",
      "  beta[i] = beta[i] * 2.0\n",
      "C:\\Users\\heinr\\anaconda3\\envs\\linkedin\\lib\\site-packages\\pyod\\models\\base.py:365: RuntimeWarning: invalid value encountered in greater\n",
      "  self.labels_ = (self.decision_scores_ > self.threshold_).astype(\n",
      "C:\\Users\\heinr\\anaconda3\\envs\\linkedin\\lib\\site-packages\\pyod\\models\\sos.py:202: RuntimeWarning: overflow encountered in multiply\n",
      "  beta[i] = beta[i] * 2.0\n",
      "C:\\Users\\heinr\\anaconda3\\envs\\linkedin\\lib\\site-packages\\pyod\\models\\base.py:365: RuntimeWarning: invalid value encountered in greater\n",
      "  self.labels_ = (self.decision_scores_ > self.threshold_).astype(\n",
      "C:\\Users\\heinr\\anaconda3\\envs\\linkedin\\lib\\site-packages\\pyod\\models\\sos.py:202: RuntimeWarning: overflow encountered in multiply\n",
      "  beta[i] = beta[i] * 2.0\n",
      "C:\\Users\\heinr\\anaconda3\\envs\\linkedin\\lib\\site-packages\\pyod\\models\\base.py:365: RuntimeWarning: invalid value encountered in greater\n",
      "  self.labels_ = (self.decision_scores_ > self.threshold_).astype(\n",
      "C:\\Users\\heinr\\anaconda3\\envs\\linkedin\\lib\\site-packages\\pyod\\models\\sos.py:202: RuntimeWarning: overflow encountered in multiply\n",
      "  beta[i] = beta[i] * 2.0\n",
      "C:\\Users\\heinr\\anaconda3\\envs\\linkedin\\lib\\site-packages\\pyod\\models\\base.py:365: RuntimeWarning: invalid value encountered in greater\n",
      "  self.labels_ = (self.decision_scores_ > self.threshold_).astype(\n",
      "C:\\Users\\heinr\\anaconda3\\envs\\linkedin\\lib\\site-packages\\pyod\\models\\sos.py:202: RuntimeWarning: overflow encountered in multiply\n",
      "  beta[i] = beta[i] * 2.0\n",
      "C:\\Users\\heinr\\anaconda3\\envs\\linkedin\\lib\\site-packages\\pyod\\models\\base.py:365: RuntimeWarning: invalid value encountered in greater\n",
      "  self.labels_ = (self.decision_scores_ > self.threshold_).astype(\n",
      "C:\\Users\\heinr\\anaconda3\\envs\\linkedin\\lib\\site-packages\\pyod\\models\\sos.py:202: RuntimeWarning: overflow encountered in multiply\n",
      "  beta[i] = beta[i] * 2.0\n",
      "C:\\Users\\heinr\\anaconda3\\envs\\linkedin\\lib\\site-packages\\pyod\\models\\base.py:365: RuntimeWarning: invalid value encountered in greater\n",
      "  self.labels_ = (self.decision_scores_ > self.threshold_).astype(\n",
      "C:\\Users\\heinr\\anaconda3\\envs\\linkedin\\lib\\site-packages\\pyod\\models\\sos.py:202: RuntimeWarning: overflow encountered in multiply\n",
      "  beta[i] = beta[i] * 2.0\n",
      "C:\\Users\\heinr\\anaconda3\\envs\\linkedin\\lib\\site-packages\\pyod\\models\\base.py:365: RuntimeWarning: invalid value encountered in greater\n",
      "  self.labels_ = (self.decision_scores_ > self.threshold_).astype(\n",
      "C:\\Users\\heinr\\anaconda3\\envs\\linkedin\\lib\\site-packages\\pyod\\models\\sos.py:202: RuntimeWarning: overflow encountered in multiply\n",
      "  beta[i] = beta[i] * 2.0\n",
      "C:\\Users\\heinr\\anaconda3\\envs\\linkedin\\lib\\site-packages\\pyod\\models\\base.py:365: RuntimeWarning: invalid value encountered in greater\n",
      "  self.labels_ = (self.decision_scores_ > self.threshold_).astype(\n",
      "C:\\Users\\heinr\\anaconda3\\envs\\linkedin\\lib\\site-packages\\pyod\\models\\sos.py:202: RuntimeWarning: overflow encountered in multiply\n",
      "  beta[i] = beta[i] * 2.0\n",
      "C:\\Users\\heinr\\anaconda3\\envs\\linkedin\\lib\\site-packages\\pyod\\models\\base.py:365: RuntimeWarning: invalid value encountered in greater\n",
      "  self.labels_ = (self.decision_scores_ > self.threshold_).astype(\n"
     ]
    }
   ],
   "source": [
    "for file in os.listdir(dir_path):\n",
    "\n",
    "    # Read file\n",
    "    file_path = os.path.join(dir_path_we, file)\n",
    "    df_temp = pd.read_csv(file_path)\n",
    "\n",
    "    # Get metadata\n",
    "    df_meta = df_temp.loc[:, 'org_profile_link':'char_length']\n",
    "\n",
    "    # Calculate company word vectors\n",
    "    df_vectors = df_temp.loc[:,'0':'299']\n",
    "    df_vector_averages = (df_vectors.sum(0) - df_vectors)/float(df_vectors.shape[0]-1)\n",
    "\n",
    "    # Calculate distances \n",
    "    df_dist = calc_distance(df_vectors, df_vector_averages)\n",
    "    \n",
    "    # Calculate anormality \n",
    "    df_anorm = calc_anormality(df_vectors)\n",
    "\n",
    "    # Merge dataframes \n",
    "    df_vectors = df_vectors.add_prefix('emp_')\n",
    "    df_vector_averages = df_vector_averages.add_prefix('comp_')\n",
    "    df_we_dist = pd.concat([df_meta, df_vectors, df_vector_averages, df_dist, df_anorm], axis=1)\n",
    "    df_we_dist_only = pd.concat([df_meta, df_dist, df_anorm], axis=1)\n",
    "    df_we_dist_only = df_we_dist_only.drop('person_summary', axis=1)\n",
    "    \n",
    "    # Export files \n",
    "    df_we_dist.to_csv(os.path.join(path_dir_dist, file))\n",
    "    df_we_dist_only.to_csv(os.path.join(path_dir_dist_only, file))\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Generate company word vectors & distance scores from LIWC"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "dir_path_liwc = os.path.normpath(r'C:\\Users\\heinr\\OneDrive\\Desktop\\LARGE DATA\\LinkedIn\\LIWC_data')\n",
    "path_dir_liwc_dist = os.path.normpath(r'C:\\Users\\heinr\\OneDrive\\Desktop\\LARGE DATA\\LinkedIn\\company_level_individual_profiles_liwc_dist')\n",
    "path_dir_liwc_dist_only = os.path.normpath(r'C:\\Users\\heinr\\OneDrive\\Desktop\\LARGE DATA\\LinkedIn\\company_level_individual_profiles_liwc_dist_only')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\heinr\\anaconda3\\envs\\linkedin\\lib\\site-packages\\pyod\\models\\sos.py:202: RuntimeWarning: overflow encountered in multiply\n",
      "  beta[i] = beta[i] * 2.0\n",
      "C:\\Users\\heinr\\anaconda3\\envs\\linkedin\\lib\\site-packages\\pyod\\models\\base.py:365: RuntimeWarning: invalid value encountered in greater\n",
      "  self.labels_ = (self.decision_scores_ > self.threshold_).astype(\n",
      "C:\\Users\\heinr\\anaconda3\\envs\\linkedin\\lib\\site-packages\\pyod\\models\\sos.py:202: RuntimeWarning: overflow encountered in multiply\n",
      "  beta[i] = beta[i] * 2.0\n",
      "C:\\Users\\heinr\\anaconda3\\envs\\linkedin\\lib\\site-packages\\pyod\\models\\base.py:365: RuntimeWarning: invalid value encountered in greater\n",
      "  self.labels_ = (self.decision_scores_ > self.threshold_).astype(\n",
      "C:\\Users\\heinr\\anaconda3\\envs\\linkedin\\lib\\site-packages\\pyod\\models\\sos.py:202: RuntimeWarning: overflow encountered in multiply\n",
      "  beta[i] = beta[i] * 2.0\n",
      "C:\\Users\\heinr\\anaconda3\\envs\\linkedin\\lib\\site-packages\\pyod\\models\\base.py:365: RuntimeWarning: invalid value encountered in greater\n",
      "  self.labels_ = (self.decision_scores_ > self.threshold_).astype(\n",
      "C:\\Users\\heinr\\anaconda3\\envs\\linkedin\\lib\\site-packages\\pyod\\models\\sos.py:202: RuntimeWarning: overflow encountered in multiply\n",
      "  beta[i] = beta[i] * 2.0\n",
      "C:\\Users\\heinr\\anaconda3\\envs\\linkedin\\lib\\site-packages\\pyod\\models\\sos.py:236: RuntimeWarning: invalid value encountered in true_divide\n",
      "  B = A / A.sum(axis=1)[:, np.newaxis]\n",
      "C:\\Users\\heinr\\anaconda3\\envs\\linkedin\\lib\\site-packages\\pyod\\models\\base.py:365: RuntimeWarning: invalid value encountered in greater\n",
      "  self.labels_ = (self.decision_scores_ > self.threshold_).astype(\n",
      "C:\\Users\\heinr\\anaconda3\\envs\\linkedin\\lib\\site-packages\\pyod\\models\\sos.py:202: RuntimeWarning: overflow encountered in multiply\n",
      "  beta[i] = beta[i] * 2.0\n",
      "C:\\Users\\heinr\\anaconda3\\envs\\linkedin\\lib\\site-packages\\pyod\\models\\base.py:365: RuntimeWarning: invalid value encountered in greater\n",
      "  self.labels_ = (self.decision_scores_ > self.threshold_).astype(\n",
      "C:\\Users\\heinr\\anaconda3\\envs\\linkedin\\lib\\site-packages\\pyod\\models\\sos.py:202: RuntimeWarning: overflow encountered in multiply\n",
      "  beta[i] = beta[i] * 2.0\n",
      "C:\\Users\\heinr\\anaconda3\\envs\\linkedin\\lib\\site-packages\\pyod\\models\\base.py:365: RuntimeWarning: invalid value encountered in greater\n",
      "  self.labels_ = (self.decision_scores_ > self.threshold_).astype(\n",
      "C:\\Users\\heinr\\anaconda3\\envs\\linkedin\\lib\\site-packages\\pyod\\models\\sos.py:202: RuntimeWarning: overflow encountered in multiply\n",
      "  beta[i] = beta[i] * 2.0\n",
      "C:\\Users\\heinr\\anaconda3\\envs\\linkedin\\lib\\site-packages\\pyod\\models\\base.py:365: RuntimeWarning: invalid value encountered in greater\n",
      "  self.labels_ = (self.decision_scores_ > self.threshold_).astype(\n",
      "C:\\Users\\heinr\\anaconda3\\envs\\linkedin\\lib\\site-packages\\pyod\\models\\sos.py:202: RuntimeWarning: overflow encountered in multiply\n",
      "  beta[i] = beta[i] * 2.0\n",
      "C:\\Users\\heinr\\anaconda3\\envs\\linkedin\\lib\\site-packages\\pyod\\models\\sos.py:236: RuntimeWarning: invalid value encountered in true_divide\n",
      "  B = A / A.sum(axis=1)[:, np.newaxis]\n",
      "C:\\Users\\heinr\\anaconda3\\envs\\linkedin\\lib\\site-packages\\pyod\\models\\base.py:365: RuntimeWarning: invalid value encountered in greater\n",
      "  self.labels_ = (self.decision_scores_ > self.threshold_).astype(\n",
      "C:\\Users\\heinr\\anaconda3\\envs\\linkedin\\lib\\site-packages\\pyod\\models\\sos.py:202: RuntimeWarning: overflow encountered in multiply\n",
      "  beta[i] = beta[i] * 2.0\n",
      "C:\\Users\\heinr\\anaconda3\\envs\\linkedin\\lib\\site-packages\\pyod\\models\\base.py:365: RuntimeWarning: invalid value encountered in greater\n",
      "  self.labels_ = (self.decision_scores_ > self.threshold_).astype(\n",
      "C:\\Users\\heinr\\anaconda3\\envs\\linkedin\\lib\\site-packages\\pyod\\models\\sos.py:202: RuntimeWarning: overflow encountered in multiply\n",
      "  beta[i] = beta[i] * 2.0\n",
      "C:\\Users\\heinr\\anaconda3\\envs\\linkedin\\lib\\site-packages\\pyod\\models\\base.py:365: RuntimeWarning: invalid value encountered in greater\n",
      "  self.labels_ = (self.decision_scores_ > self.threshold_).astype(\n",
      "C:\\Users\\heinr\\anaconda3\\envs\\linkedin\\lib\\site-packages\\pyod\\models\\sos.py:202: RuntimeWarning: overflow encountered in multiply\n",
      "  beta[i] = beta[i] * 2.0\n",
      "C:\\Users\\heinr\\anaconda3\\envs\\linkedin\\lib\\site-packages\\pyod\\models\\base.py:365: RuntimeWarning: invalid value encountered in greater\n",
      "  self.labels_ = (self.decision_scores_ > self.threshold_).astype(\n",
      "C:\\Users\\heinr\\anaconda3\\envs\\linkedin\\lib\\site-packages\\pyod\\models\\sos.py:202: RuntimeWarning: overflow encountered in multiply\n",
      "  beta[i] = beta[i] * 2.0\n",
      "C:\\Users\\heinr\\anaconda3\\envs\\linkedin\\lib\\site-packages\\pyod\\models\\base.py:365: RuntimeWarning: invalid value encountered in greater\n",
      "  self.labels_ = (self.decision_scores_ > self.threshold_).astype(\n",
      "C:\\Users\\heinr\\anaconda3\\envs\\linkedin\\lib\\site-packages\\pyod\\models\\sos.py:202: RuntimeWarning: overflow encountered in multiply\n",
      "  beta[i] = beta[i] * 2.0\n",
      "C:\\Users\\heinr\\anaconda3\\envs\\linkedin\\lib\\site-packages\\pyod\\models\\base.py:365: RuntimeWarning: invalid value encountered in greater\n",
      "  self.labels_ = (self.decision_scores_ > self.threshold_).astype(\n",
      "C:\\Users\\heinr\\anaconda3\\envs\\linkedin\\lib\\site-packages\\pyod\\models\\sos.py:202: RuntimeWarning: overflow encountered in multiply\n",
      "  beta[i] = beta[i] * 2.0\n",
      "C:\\Users\\heinr\\anaconda3\\envs\\linkedin\\lib\\site-packages\\pyod\\models\\base.py:365: RuntimeWarning: invalid value encountered in greater\n",
      "  self.labels_ = (self.decision_scores_ > self.threshold_).astype(\n",
      "C:\\Users\\heinr\\anaconda3\\envs\\linkedin\\lib\\site-packages\\pyod\\models\\sos.py:202: RuntimeWarning: overflow encountered in multiply\n",
      "  beta[i] = beta[i] * 2.0\n",
      "C:\\Users\\heinr\\anaconda3\\envs\\linkedin\\lib\\site-packages\\pyod\\models\\base.py:365: RuntimeWarning: invalid value encountered in greater\n",
      "  self.labels_ = (self.decision_scores_ > self.threshold_).astype(\n",
      "C:\\Users\\heinr\\anaconda3\\envs\\linkedin\\lib\\site-packages\\pyod\\models\\sos.py:202: RuntimeWarning: overflow encountered in multiply\n",
      "  beta[i] = beta[i] * 2.0\n",
      "C:\\Users\\heinr\\anaconda3\\envs\\linkedin\\lib\\site-packages\\pyod\\models\\base.py:365: RuntimeWarning: invalid value encountered in greater\n",
      "  self.labels_ = (self.decision_scores_ > self.threshold_).astype(\n",
      "C:\\Users\\heinr\\anaconda3\\envs\\linkedin\\lib\\site-packages\\pyod\\models\\sos.py:202: RuntimeWarning: overflow encountered in multiply\n",
      "  beta[i] = beta[i] * 2.0\n",
      "C:\\Users\\heinr\\anaconda3\\envs\\linkedin\\lib\\site-packages\\pyod\\models\\base.py:365: RuntimeWarning: invalid value encountered in greater\n",
      "  self.labels_ = (self.decision_scores_ > self.threshold_).astype(\n",
      "C:\\Users\\heinr\\anaconda3\\envs\\linkedin\\lib\\site-packages\\pyod\\models\\sos.py:202: RuntimeWarning: overflow encountered in multiply\n",
      "  beta[i] = beta[i] * 2.0\n",
      "C:\\Users\\heinr\\anaconda3\\envs\\linkedin\\lib\\site-packages\\pyod\\models\\base.py:365: RuntimeWarning: invalid value encountered in greater\n",
      "  self.labels_ = (self.decision_scores_ > self.threshold_).astype(\n",
      "C:\\Users\\heinr\\anaconda3\\envs\\linkedin\\lib\\site-packages\\pyod\\models\\sos.py:202: RuntimeWarning: overflow encountered in multiply\n",
      "  beta[i] = beta[i] * 2.0\n",
      "C:\\Users\\heinr\\anaconda3\\envs\\linkedin\\lib\\site-packages\\pyod\\models\\base.py:365: RuntimeWarning: invalid value encountered in greater\n",
      "  self.labels_ = (self.decision_scores_ > self.threshold_).astype(\n",
      "C:\\Users\\heinr\\anaconda3\\envs\\linkedin\\lib\\site-packages\\pyod\\models\\sos.py:202: RuntimeWarning: overflow encountered in multiply\n",
      "  beta[i] = beta[i] * 2.0\n",
      "C:\\Users\\heinr\\anaconda3\\envs\\linkedin\\lib\\site-packages\\pyod\\models\\base.py:365: RuntimeWarning: invalid value encountered in greater\n",
      "  self.labels_ = (self.decision_scores_ > self.threshold_).astype(\n",
      "C:\\Users\\heinr\\anaconda3\\envs\\linkedin\\lib\\site-packages\\pyod\\models\\sos.py:202: RuntimeWarning: overflow encountered in multiply\n",
      "  beta[i] = beta[i] * 2.0\n",
      "C:\\Users\\heinr\\anaconda3\\envs\\linkedin\\lib\\site-packages\\pyod\\models\\base.py:365: RuntimeWarning: invalid value encountered in greater\n",
      "  self.labels_ = (self.decision_scores_ > self.threshold_).astype(\n",
      "C:\\Users\\heinr\\anaconda3\\envs\\linkedin\\lib\\site-packages\\pyod\\models\\sos.py:202: RuntimeWarning: overflow encountered in multiply\n",
      "  beta[i] = beta[i] * 2.0\n",
      "C:\\Users\\heinr\\anaconda3\\envs\\linkedin\\lib\\site-packages\\pyod\\models\\base.py:365: RuntimeWarning: invalid value encountered in greater\n",
      "  self.labels_ = (self.decision_scores_ > self.threshold_).astype(\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\heinr\\anaconda3\\envs\\linkedin\\lib\\site-packages\\pyod\\models\\sos.py:202: RuntimeWarning: overflow encountered in multiply\n",
      "  beta[i] = beta[i] * 2.0\n",
      "C:\\Users\\heinr\\anaconda3\\envs\\linkedin\\lib\\site-packages\\pyod\\models\\base.py:365: RuntimeWarning: invalid value encountered in greater\n",
      "  self.labels_ = (self.decision_scores_ > self.threshold_).astype(\n",
      "C:\\Users\\heinr\\anaconda3\\envs\\linkedin\\lib\\site-packages\\pyod\\models\\sos.py:202: RuntimeWarning: overflow encountered in multiply\n",
      "  beta[i] = beta[i] * 2.0\n",
      "C:\\Users\\heinr\\anaconda3\\envs\\linkedin\\lib\\site-packages\\pyod\\models\\base.py:365: RuntimeWarning: invalid value encountered in greater\n",
      "  self.labels_ = (self.decision_scores_ > self.threshold_).astype(\n",
      "C:\\Users\\heinr\\anaconda3\\envs\\linkedin\\lib\\site-packages\\pyod\\models\\sos.py:202: RuntimeWarning: overflow encountered in multiply\n",
      "  beta[i] = beta[i] * 2.0\n",
      "C:\\Users\\heinr\\anaconda3\\envs\\linkedin\\lib\\site-packages\\pyod\\models\\base.py:365: RuntimeWarning: invalid value encountered in greater\n",
      "  self.labels_ = (self.decision_scores_ > self.threshold_).astype(\n",
      "C:\\Users\\heinr\\anaconda3\\envs\\linkedin\\lib\\site-packages\\pyod\\models\\sos.py:202: RuntimeWarning: overflow encountered in multiply\n",
      "  beta[i] = beta[i] * 2.0\n",
      "C:\\Users\\heinr\\anaconda3\\envs\\linkedin\\lib\\site-packages\\pyod\\models\\base.py:365: RuntimeWarning: invalid value encountered in greater\n",
      "  self.labels_ = (self.decision_scores_ > self.threshold_).astype(\n",
      "C:\\Users\\heinr\\anaconda3\\envs\\linkedin\\lib\\site-packages\\pyod\\models\\sos.py:202: RuntimeWarning: overflow encountered in multiply\n",
      "  beta[i] = beta[i] * 2.0\n",
      "C:\\Users\\heinr\\anaconda3\\envs\\linkedin\\lib\\site-packages\\pyod\\models\\base.py:365: RuntimeWarning: invalid value encountered in greater\n",
      "  self.labels_ = (self.decision_scores_ > self.threshold_).astype(\n",
      "C:\\Users\\heinr\\anaconda3\\envs\\linkedin\\lib\\site-packages\\pyod\\models\\sos.py:202: RuntimeWarning: overflow encountered in multiply\n",
      "  beta[i] = beta[i] * 2.0\n",
      "C:\\Users\\heinr\\anaconda3\\envs\\linkedin\\lib\\site-packages\\pyod\\models\\base.py:365: RuntimeWarning: invalid value encountered in greater\n",
      "  self.labels_ = (self.decision_scores_ > self.threshold_).astype(\n",
      "C:\\Users\\heinr\\anaconda3\\envs\\linkedin\\lib\\site-packages\\pyod\\models\\sos.py:202: RuntimeWarning: overflow encountered in multiply\n",
      "  beta[i] = beta[i] * 2.0\n",
      "C:\\Users\\heinr\\anaconda3\\envs\\linkedin\\lib\\site-packages\\pyod\\models\\base.py:365: RuntimeWarning: invalid value encountered in greater\n",
      "  self.labels_ = (self.decision_scores_ > self.threshold_).astype(\n",
      "C:\\Users\\heinr\\anaconda3\\envs\\linkedin\\lib\\site-packages\\pyod\\models\\sos.py:202: RuntimeWarning: overflow encountered in multiply\n",
      "  beta[i] = beta[i] * 2.0\n",
      "C:\\Users\\heinr\\anaconda3\\envs\\linkedin\\lib\\site-packages\\pyod\\models\\base.py:365: RuntimeWarning: invalid value encountered in greater\n",
      "  self.labels_ = (self.decision_scores_ > self.threshold_).astype(\n",
      "C:\\Users\\heinr\\anaconda3\\envs\\linkedin\\lib\\site-packages\\pyod\\models\\sos.py:202: RuntimeWarning: overflow encountered in multiply\n",
      "  beta[i] = beta[i] * 2.0\n",
      "C:\\Users\\heinr\\anaconda3\\envs\\linkedin\\lib\\site-packages\\pyod\\models\\base.py:365: RuntimeWarning: invalid value encountered in greater\n",
      "  self.labels_ = (self.decision_scores_ > self.threshold_).astype(\n",
      "C:\\Users\\heinr\\anaconda3\\envs\\linkedin\\lib\\site-packages\\pyod\\models\\sos.py:202: RuntimeWarning: overflow encountered in multiply\n",
      "  beta[i] = beta[i] * 2.0\n",
      "C:\\Users\\heinr\\anaconda3\\envs\\linkedin\\lib\\site-packages\\pyod\\models\\base.py:365: RuntimeWarning: invalid value encountered in greater\n",
      "  self.labels_ = (self.decision_scores_ > self.threshold_).astype(\n",
      "C:\\Users\\heinr\\anaconda3\\envs\\linkedin\\lib\\site-packages\\pyod\\models\\sos.py:202: RuntimeWarning: overflow encountered in multiply\n",
      "  beta[i] = beta[i] * 2.0\n",
      "C:\\Users\\heinr\\anaconda3\\envs\\linkedin\\lib\\site-packages\\pyod\\models\\base.py:365: RuntimeWarning: invalid value encountered in greater\n",
      "  self.labels_ = (self.decision_scores_ > self.threshold_).astype(\n",
      "C:\\Users\\heinr\\anaconda3\\envs\\linkedin\\lib\\site-packages\\pyod\\models\\sos.py:202: RuntimeWarning: overflow encountered in multiply\n",
      "  beta[i] = beta[i] * 2.0\n",
      "C:\\Users\\heinr\\anaconda3\\envs\\linkedin\\lib\\site-packages\\pyod\\models\\base.py:365: RuntimeWarning: invalid value encountered in greater\n",
      "  self.labels_ = (self.decision_scores_ > self.threshold_).astype(\n",
      "C:\\Users\\heinr\\anaconda3\\envs\\linkedin\\lib\\site-packages\\pyod\\models\\sos.py:202: RuntimeWarning: overflow encountered in multiply\n",
      "  beta[i] = beta[i] * 2.0\n",
      "C:\\Users\\heinr\\anaconda3\\envs\\linkedin\\lib\\site-packages\\pyod\\models\\base.py:365: RuntimeWarning: invalid value encountered in greater\n",
      "  self.labels_ = (self.decision_scores_ > self.threshold_).astype(\n",
      "C:\\Users\\heinr\\anaconda3\\envs\\linkedin\\lib\\site-packages\\pyod\\models\\sos.py:202: RuntimeWarning: overflow encountered in multiply\n",
      "  beta[i] = beta[i] * 2.0\n",
      "C:\\Users\\heinr\\anaconda3\\envs\\linkedin\\lib\\site-packages\\pyod\\models\\base.py:365: RuntimeWarning: invalid value encountered in greater\n",
      "  self.labels_ = (self.decision_scores_ > self.threshold_).astype(\n",
      "C:\\Users\\heinr\\anaconda3\\envs\\linkedin\\lib\\site-packages\\pyod\\models\\sos.py:202: RuntimeWarning: overflow encountered in multiply\n",
      "  beta[i] = beta[i] * 2.0\n",
      "C:\\Users\\heinr\\anaconda3\\envs\\linkedin\\lib\\site-packages\\pyod\\models\\base.py:365: RuntimeWarning: invalid value encountered in greater\n",
      "  self.labels_ = (self.decision_scores_ > self.threshold_).astype(\n",
      "C:\\Users\\heinr\\anaconda3\\envs\\linkedin\\lib\\site-packages\\pyod\\models\\sos.py:202: RuntimeWarning: overflow encountered in multiply\n",
      "  beta[i] = beta[i] * 2.0\n",
      "C:\\Users\\heinr\\anaconda3\\envs\\linkedin\\lib\\site-packages\\pyod\\models\\base.py:365: RuntimeWarning: invalid value encountered in greater\n",
      "  self.labels_ = (self.decision_scores_ > self.threshold_).astype(\n",
      "C:\\Users\\heinr\\anaconda3\\envs\\linkedin\\lib\\site-packages\\pyod\\models\\sos.py:202: RuntimeWarning: overflow encountered in multiply\n",
      "  beta[i] = beta[i] * 2.0\n",
      "C:\\Users\\heinr\\anaconda3\\envs\\linkedin\\lib\\site-packages\\pyod\\models\\base.py:365: RuntimeWarning: invalid value encountered in greater\n",
      "  self.labels_ = (self.decision_scores_ > self.threshold_).astype(\n",
      "C:\\Users\\heinr\\anaconda3\\envs\\linkedin\\lib\\site-packages\\pyod\\models\\sos.py:202: RuntimeWarning: overflow encountered in multiply\n",
      "  beta[i] = beta[i] * 2.0\n",
      "C:\\Users\\heinr\\anaconda3\\envs\\linkedin\\lib\\site-packages\\pyod\\models\\base.py:365: RuntimeWarning: invalid value encountered in greater\n",
      "  self.labels_ = (self.decision_scores_ > self.threshold_).astype(\n",
      "C:\\Users\\heinr\\anaconda3\\envs\\linkedin\\lib\\site-packages\\pyod\\models\\sos.py:202: RuntimeWarning: overflow encountered in multiply\n",
      "  beta[i] = beta[i] * 2.0\n",
      "C:\\Users\\heinr\\anaconda3\\envs\\linkedin\\lib\\site-packages\\pyod\\models\\base.py:365: RuntimeWarning: invalid value encountered in greater\n",
      "  self.labels_ = (self.decision_scores_ > self.threshold_).astype(\n",
      "C:\\Users\\heinr\\anaconda3\\envs\\linkedin\\lib\\site-packages\\pyod\\models\\sos.py:202: RuntimeWarning: overflow encountered in multiply\n",
      "  beta[i] = beta[i] * 2.0\n",
      "C:\\Users\\heinr\\anaconda3\\envs\\linkedin\\lib\\site-packages\\pyod\\models\\base.py:365: RuntimeWarning: invalid value encountered in greater\n",
      "  self.labels_ = (self.decision_scores_ > self.threshold_).astype(\n",
      "C:\\Users\\heinr\\anaconda3\\envs\\linkedin\\lib\\site-packages\\pyod\\models\\sos.py:202: RuntimeWarning: overflow encountered in multiply\n",
      "  beta[i] = beta[i] * 2.0\n",
      "C:\\Users\\heinr\\anaconda3\\envs\\linkedin\\lib\\site-packages\\pyod\\models\\base.py:365: RuntimeWarning: invalid value encountered in greater\n",
      "  self.labels_ = (self.decision_scores_ > self.threshold_).astype(\n",
      "C:\\Users\\heinr\\anaconda3\\envs\\linkedin\\lib\\site-packages\\pyod\\models\\sos.py:202: RuntimeWarning: overflow encountered in multiply\n",
      "  beta[i] = beta[i] * 2.0\n",
      "C:\\Users\\heinr\\anaconda3\\envs\\linkedin\\lib\\site-packages\\pyod\\models\\base.py:365: RuntimeWarning: invalid value encountered in greater\n",
      "  self.labels_ = (self.decision_scores_ > self.threshold_).astype(\n",
      "C:\\Users\\heinr\\anaconda3\\envs\\linkedin\\lib\\site-packages\\pyod\\models\\sos.py:202: RuntimeWarning: overflow encountered in multiply\n",
      "  beta[i] = beta[i] * 2.0\n",
      "C:\\Users\\heinr\\anaconda3\\envs\\linkedin\\lib\\site-packages\\pyod\\models\\base.py:365: RuntimeWarning: invalid value encountered in greater\n",
      "  self.labels_ = (self.decision_scores_ > self.threshold_).astype(\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\heinr\\anaconda3\\envs\\linkedin\\lib\\site-packages\\pyod\\models\\sos.py:202: RuntimeWarning: overflow encountered in multiply\n",
      "  beta[i] = beta[i] * 2.0\n",
      "C:\\Users\\heinr\\anaconda3\\envs\\linkedin\\lib\\site-packages\\pyod\\models\\base.py:365: RuntimeWarning: invalid value encountered in greater\n",
      "  self.labels_ = (self.decision_scores_ > self.threshold_).astype(\n",
      "C:\\Users\\heinr\\anaconda3\\envs\\linkedin\\lib\\site-packages\\pyod\\models\\sos.py:202: RuntimeWarning: overflow encountered in multiply\n",
      "  beta[i] = beta[i] * 2.0\n",
      "C:\\Users\\heinr\\anaconda3\\envs\\linkedin\\lib\\site-packages\\pyod\\models\\base.py:365: RuntimeWarning: invalid value encountered in greater\n",
      "  self.labels_ = (self.decision_scores_ > self.threshold_).astype(\n",
      "C:\\Users\\heinr\\anaconda3\\envs\\linkedin\\lib\\site-packages\\pyod\\models\\sos.py:202: RuntimeWarning: overflow encountered in multiply\n",
      "  beta[i] = beta[i] * 2.0\n",
      "C:\\Users\\heinr\\anaconda3\\envs\\linkedin\\lib\\site-packages\\pyod\\models\\base.py:365: RuntimeWarning: invalid value encountered in greater\n",
      "  self.labels_ = (self.decision_scores_ > self.threshold_).astype(\n",
      "C:\\Users\\heinr\\anaconda3\\envs\\linkedin\\lib\\site-packages\\pyod\\models\\sos.py:202: RuntimeWarning: overflow encountered in multiply\n",
      "  beta[i] = beta[i] * 2.0\n",
      "C:\\Users\\heinr\\anaconda3\\envs\\linkedin\\lib\\site-packages\\pyod\\models\\base.py:365: RuntimeWarning: invalid value encountered in greater\n",
      "  self.labels_ = (self.decision_scores_ > self.threshold_).astype(\n",
      "C:\\Users\\heinr\\anaconda3\\envs\\linkedin\\lib\\site-packages\\pyod\\models\\sos.py:202: RuntimeWarning: overflow encountered in multiply\n",
      "  beta[i] = beta[i] * 2.0\n",
      "C:\\Users\\heinr\\anaconda3\\envs\\linkedin\\lib\\site-packages\\pyod\\models\\base.py:365: RuntimeWarning: invalid value encountered in greater\n",
      "  self.labels_ = (self.decision_scores_ > self.threshold_).astype(\n",
      "C:\\Users\\heinr\\anaconda3\\envs\\linkedin\\lib\\site-packages\\pyod\\models\\sos.py:202: RuntimeWarning: overflow encountered in multiply\n",
      "  beta[i] = beta[i] * 2.0\n",
      "C:\\Users\\heinr\\anaconda3\\envs\\linkedin\\lib\\site-packages\\pyod\\models\\base.py:365: RuntimeWarning: invalid value encountered in greater\n",
      "  self.labels_ = (self.decision_scores_ > self.threshold_).astype(\n",
      "C:\\Users\\heinr\\anaconda3\\envs\\linkedin\\lib\\site-packages\\pyod\\models\\sos.py:202: RuntimeWarning: overflow encountered in multiply\n",
      "  beta[i] = beta[i] * 2.0\n",
      "C:\\Users\\heinr\\anaconda3\\envs\\linkedin\\lib\\site-packages\\pyod\\models\\base.py:365: RuntimeWarning: invalid value encountered in greater\n",
      "  self.labels_ = (self.decision_scores_ > self.threshold_).astype(\n",
      "C:\\Users\\heinr\\anaconda3\\envs\\linkedin\\lib\\site-packages\\pyod\\models\\sos.py:202: RuntimeWarning: overflow encountered in multiply\n",
      "  beta[i] = beta[i] * 2.0\n",
      "C:\\Users\\heinr\\anaconda3\\envs\\linkedin\\lib\\site-packages\\pyod\\models\\base.py:365: RuntimeWarning: invalid value encountered in greater\n",
      "  self.labels_ = (self.decision_scores_ > self.threshold_).astype(\n",
      "C:\\Users\\heinr\\anaconda3\\envs\\linkedin\\lib\\site-packages\\pyod\\models\\sos.py:202: RuntimeWarning: overflow encountered in multiply\n",
      "  beta[i] = beta[i] * 2.0\n",
      "C:\\Users\\heinr\\anaconda3\\envs\\linkedin\\lib\\site-packages\\pyod\\models\\base.py:365: RuntimeWarning: invalid value encountered in greater\n",
      "  self.labels_ = (self.decision_scores_ > self.threshold_).astype(\n",
      "C:\\Users\\heinr\\anaconda3\\envs\\linkedin\\lib\\site-packages\\pyod\\models\\sos.py:202: RuntimeWarning: overflow encountered in multiply\n",
      "  beta[i] = beta[i] * 2.0\n",
      "C:\\Users\\heinr\\anaconda3\\envs\\linkedin\\lib\\site-packages\\pyod\\models\\base.py:365: RuntimeWarning: invalid value encountered in greater\n",
      "  self.labels_ = (self.decision_scores_ > self.threshold_).astype(\n",
      "C:\\Users\\heinr\\anaconda3\\envs\\linkedin\\lib\\site-packages\\pyod\\models\\sos.py:202: RuntimeWarning: overflow encountered in multiply\n",
      "  beta[i] = beta[i] * 2.0\n",
      "C:\\Users\\heinr\\anaconda3\\envs\\linkedin\\lib\\site-packages\\pyod\\models\\base.py:365: RuntimeWarning: invalid value encountered in greater\n",
      "  self.labels_ = (self.decision_scores_ > self.threshold_).astype(\n",
      "C:\\Users\\heinr\\anaconda3\\envs\\linkedin\\lib\\site-packages\\pyod\\models\\sos.py:202: RuntimeWarning: overflow encountered in multiply\n",
      "  beta[i] = beta[i] * 2.0\n",
      "C:\\Users\\heinr\\anaconda3\\envs\\linkedin\\lib\\site-packages\\pyod\\models\\base.py:365: RuntimeWarning: invalid value encountered in greater\n",
      "  self.labels_ = (self.decision_scores_ > self.threshold_).astype(\n",
      "C:\\Users\\heinr\\anaconda3\\envs\\linkedin\\lib\\site-packages\\pyod\\models\\sos.py:202: RuntimeWarning: overflow encountered in multiply\n",
      "  beta[i] = beta[i] * 2.0\n",
      "C:\\Users\\heinr\\anaconda3\\envs\\linkedin\\lib\\site-packages\\pyod\\models\\base.py:365: RuntimeWarning: invalid value encountered in greater\n",
      "  self.labels_ = (self.decision_scores_ > self.threshold_).astype(\n",
      "C:\\Users\\heinr\\anaconda3\\envs\\linkedin\\lib\\site-packages\\pyod\\models\\sos.py:202: RuntimeWarning: overflow encountered in multiply\n",
      "  beta[i] = beta[i] * 2.0\n",
      "C:\\Users\\heinr\\anaconda3\\envs\\linkedin\\lib\\site-packages\\pyod\\models\\base.py:365: RuntimeWarning: invalid value encountered in greater\n",
      "  self.labels_ = (self.decision_scores_ > self.threshold_).astype(\n",
      "C:\\Users\\heinr\\anaconda3\\envs\\linkedin\\lib\\site-packages\\pyod\\models\\sos.py:202: RuntimeWarning: overflow encountered in multiply\n",
      "  beta[i] = beta[i] * 2.0\n",
      "C:\\Users\\heinr\\anaconda3\\envs\\linkedin\\lib\\site-packages\\pyod\\models\\base.py:365: RuntimeWarning: invalid value encountered in greater\n",
      "  self.labels_ = (self.decision_scores_ > self.threshold_).astype(\n",
      "C:\\Users\\heinr\\anaconda3\\envs\\linkedin\\lib\\site-packages\\pyod\\models\\sos.py:202: RuntimeWarning: overflow encountered in multiply\n",
      "  beta[i] = beta[i] * 2.0\n",
      "C:\\Users\\heinr\\anaconda3\\envs\\linkedin\\lib\\site-packages\\pyod\\models\\base.py:365: RuntimeWarning: invalid value encountered in greater\n",
      "  self.labels_ = (self.decision_scores_ > self.threshold_).astype(\n",
      "C:\\Users\\heinr\\anaconda3\\envs\\linkedin\\lib\\site-packages\\pyod\\models\\sos.py:202: RuntimeWarning: overflow encountered in multiply\n",
      "  beta[i] = beta[i] * 2.0\n",
      "C:\\Users\\heinr\\anaconda3\\envs\\linkedin\\lib\\site-packages\\pyod\\models\\base.py:365: RuntimeWarning: invalid value encountered in greater\n",
      "  self.labels_ = (self.decision_scores_ > self.threshold_).astype(\n",
      "C:\\Users\\heinr\\anaconda3\\envs\\linkedin\\lib\\site-packages\\pyod\\models\\sos.py:202: RuntimeWarning: overflow encountered in multiply\n",
      "  beta[i] = beta[i] * 2.0\n",
      "C:\\Users\\heinr\\anaconda3\\envs\\linkedin\\lib\\site-packages\\pyod\\models\\base.py:365: RuntimeWarning: invalid value encountered in greater\n",
      "  self.labels_ = (self.decision_scores_ > self.threshold_).astype(\n",
      "C:\\Users\\heinr\\anaconda3\\envs\\linkedin\\lib\\site-packages\\pyod\\models\\sos.py:202: RuntimeWarning: overflow encountered in multiply\n",
      "  beta[i] = beta[i] * 2.0\n",
      "C:\\Users\\heinr\\anaconda3\\envs\\linkedin\\lib\\site-packages\\pyod\\models\\base.py:365: RuntimeWarning: invalid value encountered in greater\n",
      "  self.labels_ = (self.decision_scores_ > self.threshold_).astype(\n",
      "C:\\Users\\heinr\\anaconda3\\envs\\linkedin\\lib\\site-packages\\pyod\\models\\sos.py:202: RuntimeWarning: overflow encountered in multiply\n",
      "  beta[i] = beta[i] * 2.0\n",
      "C:\\Users\\heinr\\anaconda3\\envs\\linkedin\\lib\\site-packages\\pyod\\models\\base.py:365: RuntimeWarning: invalid value encountered in greater\n",
      "  self.labels_ = (self.decision_scores_ > self.threshold_).astype(\n",
      "C:\\Users\\heinr\\anaconda3\\envs\\linkedin\\lib\\site-packages\\pyod\\models\\sos.py:202: RuntimeWarning: overflow encountered in multiply\n",
      "  beta[i] = beta[i] * 2.0\n",
      "C:\\Users\\heinr\\anaconda3\\envs\\linkedin\\lib\\site-packages\\pyod\\models\\base.py:365: RuntimeWarning: invalid value encountered in greater\n",
      "  self.labels_ = (self.decision_scores_ > self.threshold_).astype(\n",
      "C:\\Users\\heinr\\anaconda3\\envs\\linkedin\\lib\\site-packages\\pyod\\models\\sos.py:202: RuntimeWarning: overflow encountered in multiply\n",
      "  beta[i] = beta[i] * 2.0\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "us-army\n",
      "Unable to allocate 30.7 GiB for an array with shape (64168, 64168) and data type float64\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\heinr\\anaconda3\\envs\\linkedin\\lib\\site-packages\\pyod\\models\\sos.py:202: RuntimeWarning: overflow encountered in multiply\n",
      "  beta[i] = beta[i] * 2.0\n",
      "C:\\Users\\heinr\\anaconda3\\envs\\linkedin\\lib\\site-packages\\pyod\\models\\base.py:365: RuntimeWarning: invalid value encountered in greater\n",
      "  self.labels_ = (self.decision_scores_ > self.threshold_).astype(\n",
      "C:\\Users\\heinr\\anaconda3\\envs\\linkedin\\lib\\site-packages\\pyod\\models\\sos.py:202: RuntimeWarning: overflow encountered in multiply\n",
      "  beta[i] = beta[i] * 2.0\n",
      "C:\\Users\\heinr\\anaconda3\\envs\\linkedin\\lib\\site-packages\\pyod\\models\\base.py:365: RuntimeWarning: invalid value encountered in greater\n",
      "  self.labels_ = (self.decision_scores_ > self.threshold_).astype(\n",
      "C:\\Users\\heinr\\anaconda3\\envs\\linkedin\\lib\\site-packages\\pyod\\models\\sos.py:202: RuntimeWarning: overflow encountered in multiply\n",
      "  beta[i] = beta[i] * 2.0\n",
      "C:\\Users\\heinr\\anaconda3\\envs\\linkedin\\lib\\site-packages\\pyod\\models\\base.py:365: RuntimeWarning: invalid value encountered in greater\n",
      "  self.labels_ = (self.decision_scores_ > self.threshold_).astype(\n"
     ]
    }
   ],
   "source": [
    "failed = []\n",
    "\n",
    "for comp_name in df_liwc_comp_names:\n",
    "\n",
    "    try:\n",
    "        # Cut dataset\n",
    "        df_liwc_temp = df_liwc[df_liwc.company == comp_name].reset_index()\n",
    "        df_meta = df_liwc_temp[['company', 'Filename.x', 'person_id']]\n",
    "        df_liwc_temp_emp = df_liwc_temp.loc[:,'Segment.x':'OtherP.x']\n",
    "\n",
    "        # Calculate company matrix\n",
    "        df_liwc_averages = (df_liwc_temp_emp.sum(0) - df_liwc_temp_emp)/float(df_liwc_temp_emp.shape[0]-1)\n",
    "\n",
    "        # Calculate distances \n",
    "        df_dist = calc_distance(df_liwc_temp_emp, df_liwc_averages)\n",
    "\n",
    "        # Calculate anormality \n",
    "        df_anorm = calc_anormality(df_liwc_temp_emp)\n",
    "\n",
    "        # Merge dataframes \n",
    "        df_liwc_temp_emp = df_liwc_temp_emp.add_prefix('emp_')\n",
    "        df_liwc_averages = df_liwc_averages.add_prefix('comp_')\n",
    "        df_liwc_dist = pd.concat([df_meta, df_liwc_temp_emp, df_liwc_averages, df_dist, df_anorm], axis=1)\n",
    "        df_liwc_dist_only = pd.concat([df_meta, df_dist, df_anorm], axis=1)\n",
    "\n",
    "        # Export files \n",
    "        df_liwc_dist.to_csv(os.path.join(path_dir_liwc_dist, comp_name+'.csv'))\n",
    "        df_liwc_dist_only.to_csv(os.path.join(path_dir_liwc_dist_only, comp_name+'.csv'))\n",
    "    \n",
    "    except Exception as e:\n",
    "        \n",
    "        failed.append(comp_name)\n",
    "        print(comp_name)\n",
    "        print(e)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "path_dir_liwc_diff = os.path.normpath(r'C:\\Users\\heinr\\OneDrive\\Desktop\\LARGE DATA\\LinkedIn\\company_level_individual_profiles_liwc_diff')\n",
    "path_dir_liwc_diff_only = os.path.normpath(r'C:\\Users\\heinr\\OneDrive\\Desktop\\LARGE DATA\\LinkedIn\\company_level_individual_profiles_liwc_diff_only')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "failed = []\n",
    "\n",
    "for file in os.listdir(path_dir_liwc_dist): \n",
    "    \n",
    "    try: \n",
    "        # Read file\n",
    "        file_path = os.path.join(path_dir_liwc_dist, file)\n",
    "        df_temp = pd.read_csv(file_path, index_col=0)\n",
    "\n",
    "        # Cut dataset\n",
    "        df_meta = df_temp[['company', 'Filename.x', 'person_id']]\n",
    "        df_liwc_temp_emp = df_temp.loc[:,'emp_Segment.x':'emp_OtherP.x']\n",
    "        df_liwc_temp_comp = df_temp.loc[:,'comp_Segment.x':'comp_OtherP.x']\n",
    "\n",
    "        # Calculate difference scores\n",
    "        df_diff = df_liwc_temp_emp - df_liwc_temp_comp.values\n",
    "        df_diff.columns = df_diff.columns.str.strip('emp_')\n",
    "        df_diff = df_diff.add_prefix('diff_')\n",
    "\n",
    "        # Merge data\n",
    "        df_liwc_diff = pd.concat([df_temp, df_diff], axis=1)\n",
    "        df_liwc_diff_only = pd.concat([df_meta, df_diff], axis=1)\n",
    "\n",
    "        # Export files \n",
    "        df_liwc_diff.to_csv(os.path.join(path_dir_liwc_diff, file))\n",
    "        df_liwc_diff_only.to_csv(os.path.join(path_dir_liwc_diff_only, file))\n",
    "    \n",
    "    except Exception as e:\n",
    "\n",
    "        failed.append(comp_name)\n",
    "        print(comp_name)\n",
    "        print(e)   \n",
    "    \n",
    "    "
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
